solr/core/src/java/org/apache/solr/cloud/ShardLeaderElectionContextBase.java

```diff
@@ -179,7 +183,16 @@ void runLeaderProcess(boolean weAreReplacement, int pauseBeforeStartMs)
           ZkStateReader.STATE_PROP, Replica.State.ACTIVE.toString());
       assert zkController != null;
       assert zkController.getOverseer() != null;
-      zkController.getOverseer().offerStateUpdate(Utils.toJSON(m));
+      DocCollection coll = zkStateReader.getCollection(this.collection);
+      if (coll == null || coll.getStateFormat() < 2 || ZkController.sendToOverseer(coll, id)) {
+        zkController.getOverseer().offerStateUpdate(Utils.toJSON(m));
+      } else {
+        PerReplicaStates prs = PerReplicaStates.fetch(coll.getZNode(), zkClient, coll.getPerReplicaStates());
+        PerReplicaStates.WriteOps writeOps = PerReplicaStates.WriteOps.flipLeader(zkStateReader.getClusterState().getCollection(collection).getSlice(shardId).getReplicaNames(), id, prs);
+        //nocommit make this debug
+        log.info("bypassed Zookeeper for leader election for {}/{}, old:{},new {} ", this.collection, shardId, currentLeader, id);
+        PerReplicaStates.persist(writeOps, coll.getZNode(), zkStateReader.getZkClient());
```

Comment:
This fetch/WriteOps.X/persist method flow seems very clunky to me, I've made a few comments on PerReplicaStates about how we could try to make it better. Another helpful change would be more descriptive naming of the variables instead of repeating the type name